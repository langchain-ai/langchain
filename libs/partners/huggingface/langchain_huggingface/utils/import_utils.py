import importlib.metadata
import importlib.util
import operator as op
from typing import Union

from packaging import version

STR_OPERATION_TO_FUNC = {
    ">": op.gt,
    ">=": op.ge,
    "==": op.eq,
    "!=": op.ne,
    "<=": op.le,
    "<": op.lt,
}


_optimum_available = importlib.util.find_spec("optimum") is not None
_optimum_version = "N/A"
if _optimum_available:
    try:
        _optimum_version = importlib.metadata.version("optimum")
    except importlib.metadata.PackageNotFoundError:
        _optimum_available = False


_optimum_intel_available = (
    _optimum_available and importlib.util.find_spec("optimum.intel") is not None
)
_optimum_intel_version = "N/A"
if _optimum_intel_available:
    try:
        _optimum_intel_version = importlib.metadata.version("optimum-intel")
    except importlib.metadata.PackageNotFoundError:
        _optimum_intel_available = False


_ipex_available = importlib.util.find_spec("intel_extension_for_pytorch") is not None

_openvino_available = importlib.util.find_spec("openvino") is not None


# This function was copied from: https://github.com/huggingface/accelerate/blob/874c4967d94badd24f893064cc3bef45f57cadf7/src/accelerate/utils/versions.py#L319
def compare_versions(
    library_or_version: Union[str, version.Version],
    operation: str,
    requirement_version: str,
) -> bool:
    """
    Compare a library version to some requirement using a given operation.

    Arguments:
        library_or_version (`str` or `packaging.version.Version`):
            A library name or a version to check.
        operation (`str`):
            A string representation of an operator, such as `">"` or `"<="`.
        requirement_version (`str`):
            The version to compare the library version against
    """
    if operation not in STR_OPERATION_TO_FUNC.keys():
        raise ValueError(
            f"`operation` must be one of {list(STR_OPERATION_TO_FUNC.keys())}"
            f", received {operation}"
        )
    if isinstance(library_or_version, str):
        library_or_version = version.parse(
            importlib.metadata.version(library_or_version)
        )
    return STR_OPERATION_TO_FUNC[operation](
        library_or_version, version.parse(requirement_version)
    )


def is_optimum_available() -> bool:
    return _optimum_available


def is_optimum_intel_available() -> bool:
    return _optimum_intel_available


def is_ipex_available() -> bool:
    return _ipex_available


def is_openvino_available() -> bool:
    return _openvino_available


def is_optimum_version(operation: str, reference_version: str) -> bool:
    """
    Compare the current Optimum version to a given reference with an operation.
    """
    if not _optimum_version:
        return False
    return compare_versions(
        version.parse(_optimum_version), operation, reference_version
    )


def is_optimum_intel_version(operation: str, reference_version: str) -> bool:
    """
    Compare the current Optimum Intel version to a given reference with an operation.
    """
    if not _optimum_intel_version:
        return False
    return compare_versions(
        version.parse(_optimum_intel_version), operation, reference_version
    )


IMPORT_ERROR = """
requires the {0} library but it was not found in your environment.
You can install it with pip: `pip install {0}`.
Please note that you may need to restart your runtime after installation.
"""
