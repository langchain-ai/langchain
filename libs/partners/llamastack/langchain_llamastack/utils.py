"""
Minimal utility functions for LlamaStack operations.
"""

import logging
from typing import Any, Dict, List, Optional

try:
    from llama_stack_client import LlamaStackClient
except ImportError:
    LlamaStackClient = None

logger = logging.getLogger(__name__)


def check_llamastack_connection(
    base_url: str = "http://localhost:8321",
    llamastack_api_key: Optional[str] = None,
    model_type: str = "inference",
) -> Dict[str, Any]:
    """Check LlamaStack connection and return status information.

    Args:
        base_url: LlamaStack base URL
        llamastack_api_key: LlamaStack API key (if required)
        model_type: Type of models to filter for ('inference', 'embedding', or 'all')

    Returns:
        Dictionary with connection status and model information

    Example:
        status = check_llamastack_connection()
        if status['connected']:
            print(f"Connected! {status['models_count']} models available")
            print(f"Models: {status['models']}")
        else:
            print(f"Connection failed: {status['error']}")
    """
    if LlamaStackClient is None:
        return {
            "connected": False,
            "base_url": base_url,
            "models_count": 0,
            "models": [],
            "error": "llama-stack-client not available. Install with: pip install llama-stack-client",
        }

    try:
        # Create temporary client to check models
        client_kwargs = {"base_url": base_url}
        if llamastack_api_key:
            client_kwargs["api_key"] = llamastack_api_key

        temp_client = LlamaStackClient(**client_kwargs)
        models = temp_client.models.list()

        # Filter models by type
        if model_type == "all":
            model_ids = [model.identifier for model in models]
        else:
            model_ids = []
            for model in models:
                model_id = model.identifier
                model_type_attr = getattr(model, "model_type", None)

                # Filter based on model type or common naming patterns
                if model_type == "inference":
                    # Exclude guard/shield models first (priority exclusion)
                    if any(
                        pattern in model_id.lower()
                        for pattern in [
                            "guard",
                            "shield",
                            "safety",
                            "moderation",
                            "classifier",
                        ]
                    ):
                        continue

                    # Exclude embedding models
                    if any(
                        pattern in model_id.lower()
                        for pattern in [
                            "embed",
                            "embedding",
                            "minilm",
                            "sentence",
                            "bge",
                            "e5",
                        ]
                    ):
                        continue

                    # Include models that are likely for chat/inference
                    if (
                        model_type_attr == "text_completion"
                        or model_type_attr == "chat_completion"
                        or model_type_attr
                        is None  # Default assumption for untyped models
                        or any(
                            pattern in model_id.lower()
                            for pattern in [
                                "instruct",  # Most important pattern
                                "chat",
                                "llama",
                                "mistral",
                                "codellama",
                                "vicuna",
                                "alpaca",
                                "qwen",
                                "gemma",
                                "phi",
                                "wizardlm",
                                "orca",
                                "turbo",
                            ]
                        )
                    ):
                        model_ids.append(model_id)
                elif model_type == "embedding":
                    # Include models that are for embeddings
                    if model_type_attr == "embedding" or any(
                        pattern in model_id.lower()
                        for pattern in [
                            "embed",
                            "embedding",
                            "minilm",
                            "sentence",
                            "bge",
                            "e5",
                        ]
                    ):
                        model_ids.append(model_id)

        # Sort models to prioritize most reliable ones
        def model_priority(model_id):
            """Sort key function to prioritize models."""
            # Fireworks/OpenAI/etc models first (more reliable routing)
            if not model_id.startswith("ollama/"):
                return 0
            # Instruct models within ollama
            if "instruct" in model_id.lower():
                return 1
            # Other ollama models
            return 2

        # Sort models by priority
        model_ids.sort(key=model_priority)

        logger.debug(
            f"Found {len(model_ids)} {model_type} models in LlamaStack: {model_ids}"
        )

        return {
            "connected": True,
            "base_url": base_url,
            "models_count": len(model_ids),
            "models": model_ids,
            "error": None,
        }
    except Exception as e:
        logger.error(f"Failed to connect to LlamaStack at {base_url}: {e}")
        return {
            "connected": False,
            "base_url": base_url,
            "models_count": 0,
            "models": [],
            "error": str(e),
        }


def list_available_models(
    base_url: str = "http://localhost:8321",
    llamastack_api_key: Optional[str] = None,
    model_type: str = "inference",
) -> List[str]:
    """List available models from LlamaStack.

    Args:
        base_url: LlamaStack base URL
        llamastack_api_key: LlamaStack API key (if required)
        model_type: Type of models to filter for ('inference', 'embedding', or 'all')

    Returns:
        List of available model names

    Raises:
        ValueError: If LlamaStack is not available
    """
    status = check_llamastack_connection(base_url, llamastack_api_key, model_type)

    if status["connected"]:
        return status["models"]
    else:
        raise ValueError(f"Cannot connect to LlamaStack: {status['error']}")
