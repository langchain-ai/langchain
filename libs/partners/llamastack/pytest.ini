[tool:pytest]
testpaths = tests
python_files = test_*.py
python_classes = Test*
python_functions = test_*
addopts = -v --tb=short
markers =
    unit: Unit tests
    integration: Integration tests that require a running Llama Stack server
