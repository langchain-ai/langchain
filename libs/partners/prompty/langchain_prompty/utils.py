import traceback
from pathlib import Path
from typing import Any, Union

from .core import (
    Frontmatter,
    InvokerFactory,
    ModelSettings,
    Prompty,
    PropertySettings,
    SimpleModel,
    TemplateSettings,
    param_hoisting,
)


def load(prompt_path: str, configuration: str = "default") -> Prompty:
    """Load a prompty file and return a Prompty object.

    Args:
        prompt_path: The path to the prompty file.
        configuration: The configuration to use. Defaults to "default".

    Returns:
        The Prompty object.
    """
    file_path = Path(prompt_path)
    if not file_path.is_absolute():
        # get caller's path (take into account trace frame)
        caller = Path(traceback.extract_stack()[-3].filename)
        file_path = Path(caller.parent / file_path).resolve().absolute()

    # load dictionary from prompty file
    matter = Frontmatter.read_file(file_path.__fspath__())
    attributes = matter["attributes"]
    content = matter["body"]

    # normalize attribute dictionary resolve keys and files
    attributes = Prompty.normalize(attributes, file_path.parent)

    # load global configuration
    if "model" not in attributes:
        attributes["model"] = {}

    # pull model settings out of attributes
    try:
        model = ModelSettings(**attributes.pop("model"))
    except Exception as e:
        raise ValueError(f"Error in model settings: {e}")

    # pull template settings
    try:
        if "template" in attributes:
            t = attributes.pop("template")
            if isinstance(t, dict):
                template = TemplateSettings(**t)
            # has to be a string denoting the type
            else:
                template = TemplateSettings(type=t, parser="prompty")
        else:
            template = TemplateSettings(type="mustache", parser="prompty")
    except Exception as e:
        raise ValueError(f"Error in template loader: {e}")

    # formalize inputs and outputs
    if "inputs" in attributes:
        try:
            inputs = {
                k: PropertySettings(**v) for (k, v) in attributes.pop("inputs").items()
            }
        except Exception as e:
            raise ValueError(f"Error in inputs: {e}")
    else:
        inputs = {}
    if "outputs" in attributes:
        try:
            outputs = {
                k: PropertySettings(**v) for (k, v) in attributes.pop("outputs").items()
            }
        except Exception as e:
            raise ValueError(f"Error in outputs: {e}")
    else:
        outputs = {}

    # recursive loading of base prompty
    if "base" in attributes:
        # load the base prompty from the same directory as the current prompty
        base = load(file_path.parent / attributes["base"])
        # hoist the base prompty's attributes to the current prompty
        model.api = base.model.api if model.api == "" else model.api
        model.configuration = param_hoisting(
            model.configuration, base.model.configuration
        )
        model.parameters = param_hoisting(model.parameters, base.model.parameters)
        model.response = param_hoisting(model.response, base.model.response)
        attributes["sample"] = param_hoisting(attributes, base.sample, "sample")

        p = Prompty(
            **attributes,
            model=model,
            inputs=inputs,
            outputs=outputs,
            template=template,
            content=content,
            file=file_path,
            basePrompty=base,
        )
    else:
        p = Prompty(
            **attributes,
            model=model,
            inputs=inputs,
            outputs=outputs,
            template=template,
            content=content,
            file=file_path,
        )
    return p


def prepare(
    prompt: Prompty,
    inputs: dict[str, Any] = {},
) -> Any:
    """Prepare the inputs for the prompty.

    Args:
        prompt: The Prompty object.
        inputs: The inputs to the prompty. Defaults to {}.

    Returns:
        The prepared inputs.
    """
    invoker = InvokerFactory()

    inputs = param_hoisting(inputs, prompt.sample)

    if prompt.template.type == "NOOP":
        render = prompt.content
    else:
        # render
        result = invoker(
            "renderer",
            prompt.template.type,
            prompt,
            SimpleModel(item=inputs),
        )
        render = result.item

    if prompt.template.parser == "NOOP":
        result = render
    else:
        # parse
        result = invoker(
            "parser",
            f"{prompt.template.parser}.{prompt.model.api}",
            prompt,
            SimpleModel(item=result.item),
        )

    if isinstance(result, SimpleModel):
        return result.item
    else:
        return result


def run(
    prompt: Prompty,
    content: Union[dict, list, str],
    configuration: dict[str, Any] = {},
    parameters: dict[str, Any] = {},
    raw: bool = False,
) -> Any:
    """Run the prompty.

    Args:
        prompt: The Prompty object.
        content: The content to run the prompty on.
        configuration: The configuration to use. Defaults to {}.
        parameters: The parameters to use. Defaults to {}.
        raw: Whether to return the raw output. Defaults to False.

    Returns:
        The result of running the prompty.
    """
    invoker = InvokerFactory()

    if configuration != {}:
        prompt.model.configuration = param_hoisting(
            configuration, prompt.model.configuration
        )

    if parameters != {}:
        prompt.model.parameters = param_hoisting(parameters, prompt.model.parameters)

    # execute
    result = invoker(
        "executor",
        prompt.model.configuration["type"],
        prompt,
        SimpleModel(item=content),
    )

    # skip?
    if not raw:
        # process
        result = invoker(
            "processor",
            prompt.model.configuration["type"],
            prompt,
            result,
        )

    if isinstance(result, SimpleModel):
        return result.item
    else:
        return result


def execute(
    prompt: Union[str, Prompty],
    configuration: dict[str, Any] = {},
    parameters: dict[str, Any] = {},
    inputs: dict[str, Any] = {},
    raw: bool = False,
    connection: str = "default",
) -> Any:
    """Execute a prompty.

    Args:
        prompt: The prompt to execute.
            Can be a path to a prompty file or a Prompty object.
        configuration: The configuration to use. Defaults to {}.
        parameters: The parameters to use. Defaults to {}.
        inputs: The inputs to the prompty. Defaults to {}.
        raw: Whether to return the raw output. Defaults to False.
        connection: The connection to use. Defaults to "default".

    Returns:
        The result of executing the prompty.
    """

    if isinstance(prompt, str):
        prompt = load(prompt, connection)

    # prepare content
    content = prepare(prompt, inputs)

    # run LLM model
    result = run(prompt, content, configuration, parameters, raw)

    return result
