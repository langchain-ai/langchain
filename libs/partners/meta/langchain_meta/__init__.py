"""This package provides the LlamaAPI integration for LangChain."""

from langchain_meta.chat_models import ChatLlama

__all__ = ["ChatLlama"]
