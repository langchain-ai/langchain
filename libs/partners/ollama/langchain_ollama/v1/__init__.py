from langchain_ollama.v1.chat_models import (
    ChatOllama,
)

__all__ = ["ChatOllama"]
