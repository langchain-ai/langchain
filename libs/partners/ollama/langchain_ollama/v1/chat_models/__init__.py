from ollama import AsyncClient, Client

from langchain_ollama._utils import validate_model
from langchain_ollama.v1.chat_models.base import (
    ChatOllama,
    _parse_arguments_from_tool_call,
    _parse_json_string,
)

__all__ = [
    "AsyncClient",
    "ChatOllama",
    "Client",
    "_parse_arguments_from_tool_call",
    "_parse_json_string",
    "validate_model",
]
