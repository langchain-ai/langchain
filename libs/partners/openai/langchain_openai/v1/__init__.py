from langchain_openai.v1.chat_models import ChatOpenAI

__all__ = ["ChatOpenAI"]
