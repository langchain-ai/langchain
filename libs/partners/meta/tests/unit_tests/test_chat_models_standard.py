"""Standard LangChain interface tests"""

from typing import Any

import pytest
from langchain_core.language_models import BaseChatModel
from langchain_tests.unit_tests import (  # type: ignore[import-not-found]
    ChatModelUnitTests,
)
from langchain_tests.utils.pydantic import PYDANTIC_MAJOR_VERSION
from pydantic import BaseModel, Field
from pydantic.v1 import BaseModel as BaseModelV1
from pydantic.v1 import Field as FieldV1

from langchain_meta import ChatLlama


def generate_schema_pydantic_v1_from_2() -> Any:
    """
    Use to generate a schema from v1 namespace in pydantic 2.

    :private:
    """
    if PYDANTIC_MAJOR_VERSION != 2:
        raise AssertionError("This function is only compatible with Pydantic v2.")

    class PersonB(BaseModelV1):
        """Record attributes of a person."""

        name: str = FieldV1(..., description="The name of the person.")
        age: int = FieldV1(..., description="The age of the person.")

    return PersonB


def generate_schema_pydantic() -> Any:
    """
    Works with either pydantic 1 or 2

    :private:
    """

    class PersonA(BaseModel):
        """Record attributes of a person."""

        name: str = Field(..., description="The name of the person.")
        age: int = Field(..., description="The age of the person.")

    return PersonA


TEST_PYDANTIC_MODELS = [generate_schema_pydantic()]

if PYDANTIC_MAJOR_VERSION == 2:
    TEST_PYDANTIC_MODELS.append(generate_schema_pydantic_v1_from_2())


class TestLlamaStandard(ChatModelUnitTests):
    @property
    def chat_model_class(self) -> type[BaseChatModel]:
        return ChatLlama

    @property
    def chat_model_params(self) -> dict:
        return {"model": "Llama-3.3-8B-Instruct"}

    @property
    def init_from_env_params(self) -> tuple[dict, dict, dict]:
        return (
            {
                "LLAMA_API_KEY": "api_key",
            },
            {
                "model": "Llama-3.3-8B-Instruct",
            },
            {
                "llama_api_key": "api_key",
                "llama_api_base": "https://api.llama.com/compat/v1/",
            },
        )

    @pytest.mark.parametrize("schema", TEST_PYDANTIC_MODELS)
    @pytest.mark.xfail(
        reason="Llama models only support 'json_schema' method,"
        "other methods are expected to fail."
    )
    def test_with_structured_output(
        self,
        model: BaseChatModel,
        schema: Any,
    ) -> None:
        """
        Test that Llama models only support
        the 'json_schema' method for structured output.
        """
        if not self.has_structured_output:
            return

        # Test default method works (should be json_schema)
        assert model.with_structured_output(schema) is not None

        # Test json_schema method explicitly
        for strict in [None, False, True]:
            assert model.with_structured_output(
                schema, method="json_schema", strict=strict
            )

        # Test that other methods raise ValueError
        for method in ["function_calling", "json_mode"]:
            with pytest.raises(
                ValueError, match="Only 'json_schema' method is supported"
            ):
                model.with_structured_output(schema, method=method)
