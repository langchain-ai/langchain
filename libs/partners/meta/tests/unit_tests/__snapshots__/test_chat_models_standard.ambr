# serializer version: 1
# name: TestLlamaStandard.test_serdes[serialized]
  dict({
    'id': list([
      'langchain_meta',
      'chat_models',
      'ChatLlama',
    ]),
    'kwargs': dict({
      'llama_api_base': 'https://api.llama.com/compat/v1/',
      'llama_api_key': dict({
        'id': list([
          'LLAMA_API_KEY',
        ]),
        'lc': 1,
        'type': 'secret',
      }),
      'max_retries': 2,
      'max_tokens': 100,
      'model_name': 'Llama-3.3-8B-Instruct',
      'request_timeout': 60.0,
      'stop': list([
      ]),
      'temperature': 0.0,
    }),
    'lc': 1,
    'name': 'ChatLlama',
    'type': 'constructor',
  })
# ---
